core/src/main/java/org/mule/runtime/core/internal/policy/PolicyTraceLogger.java

```diff
@@ -0,0 +1,126 @@
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.policy;
+
+import static org.mule.runtime.core.privileged.event.PrivilegedEvent.setCurrentEvent;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import org.mule.runtime.api.component.Component;
+import org.mule.runtime.api.metadata.TypedValue;
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.core.api.policy.Policy;
+import org.mule.runtime.core.privileged.event.PrivilegedEvent;
+
+import java.util.Map;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+
+public class PolicyTraceLogger {
+
+  private static final Logger LOGGER = getLogger(PolicyTraceLogger.class);
+  private static final String NL = System.lineSeparator();
+  private static final String TAB = "   ";
+
+  /**
+   * Logs the event before starting to execute the source policy
+   */
+  public void logSourcePolicyStart(Policy policy, CoreEvent event) {
+    if (LOGGER.isTraceEnabled()) {
+      // Setting event in thread local so the correlation Id is printed with the log message
+      setCurrentEvent((PrivilegedEvent) event);
+      LOGGER.trace(NL + "Executing policy " + getPolicyName(policy) + NL + eventAsString(event));
+    }
+  }
+
+  /**
+   * Logs the event before exiting current policy because of the execution of an execute-next element
+   */
+  public void logBeforeExecuteNext(String policyId, CoreEvent event) {
+    logEvent(() -> NL + "Jumping to next from policy " + policyId, event);
+  }
+
+  /**
+   * Logs the event after execute-next is completed and resuming execution of a policy
+   */
+  public void logAfterExecuteNext(String policyId, CoreEvent event) {
+    logEvent(() -> NL + "Resuming execution of policy " + policyId, event);
+  }
+
+  /**
+   * Logs the event before starting to execute the operation policy
+   */
+  public void logOperationPolicyStart(Policy policy, CoreEvent event) {
+    logEvent(() -> NL + "Executing operation policy " + getPolicyName(policy), event);
+  }
+
+  /**
+   * Logs the event after execution of the source policy finished
+   */
+  public void logSourcePolicyEnd(Policy policy, CoreEvent event) {
+    logEvent(() -> NL + "Policy " + getPolicyName(policy) + " execution finished", event);
+  }
+
+  /**
+   * Logs the event after execution of the operation policy finished
+   */
+  public void logOperationPolicyEnd(Policy policy, CoreEvent event) {
+    logEvent(() -> NL + "Operation policy " + getPolicyName(policy) + " execution finished", event);
+  }
+
+  public void logSourcePolicyFailureResult(SourcePolicyFailureResult result) {
+    if (LOGGER.isTraceEnabled()) {
+      LOGGER.trace("Event id: " + result.getMessagingException().getEvent().getContext().getId()
+          + "\nFinished processing with failure. \n" +
+          "Error message: " + result.getMessagingException().getMessage());
+    }
+  }
+
+  private void logEvent(Supplier<String> initialLine, CoreEvent event) {
+    if (LOGGER.isTraceEnabled()) {
```

Comment:
ask for this in the caller method, to avoid instantiating the lambda with the message